In [ ]:
import numpy as np
import pandas as pd

In [ ]:
#!pip install feature-engine

In [ ]:
df=pd.read_csv("/content/housing.csv")

In [ ]:
df.head()

longitude  latitude  ...  median_house_value  ocean_proximity
0    -122.23     37.88  ...            452600.0         NEAR BAY
1    -122.22     37.86  ...            358500.0         NEAR BAY
2    -122.24     37.85  ...            352100.0         NEAR BAY
3    -122.25     37.85  ...            341300.0         NEAR BAY
4    -122.25     37.85  ...            342200.0         NEAR BAY

[5 rows x 10 columns]

In [ ]:
df.info(memory_usage="deep")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 2.7 MB


In [ ]:
df.describe()

longitude      latitude  ...  median_income  median_house_value
count  20640.000000  20640.000000  ...   20640.000000        20640.000000
mean    -119.569704     35.631861  ...       3.870671       206855.816909
std        2.003532      2.135952  ...       1.899822       115395.615874
min     -124.350000     32.540000  ...       0.499900        14999.000000
25%     -121.800000     33.930000  ...       2.563400       119600.000000
50%     -118.490000     34.260000  ...       3.534800       179700.000000
75%     -118.010000     37.710000  ...       4.743250       264725.000000
max     -114.310000     41.950000  ...      15.000100       500001.000000

[8 rows x 9 columns]

scaling is required
****
ALso total rooms and bedrooms are way high for block.......it is given a block in range 600,3000

In [ ]:
df.isnull().mean()

longitude             0.000000
latitude              0.000000
housing_median_age    0.000000
total_rooms           0.000000
total_bedrooms        0.010029
population            0.000000
households            0.000000
median_income         0.000000
median_house_value    0.000000
ocean_proximity       0.000000
dtype: float64

Since number of null is near 1% and entries are 20640 we can drop null values

In [ ]:
df["ocean_proximity"].value_counts()

<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: ocean_proximity, dtype: int64

In [ ]:
df.corr()

longitude  latitude  ...  median_income  median_house_value
longitude            1.000000 -0.924664  ...      -0.015176           -0.045967
latitude            -0.924664  1.000000  ...      -0.079809           -0.144160
housing_median_age  -0.108197  0.011173  ...      -0.119034            0.105623
total_rooms          0.044568 -0.036100  ...       0.198050            0.134153
total_bedrooms       0.069608 -0.066983  ...      -0.007723            0.049686
population           0.099773 -0.108785  ...       0.004834           -0.024650
households           0.055310 -0.071035  ...       0.013033            0.065843
median_income       -0.015176 -0.079809  ...       1.000000            0.688075
median_house_value  -0.045967 -0.144160  ...       0.688075            1.000000

[9 rows x 9 columns]

*   **We can see median_house_value depends highly on median_income which make sense**
*    **Among total_rooms,total_bedrooms,population and households 
I have choosen *total_rooms as it has higher corr with price***
also took latitude 

In [ ]:
df=df.dropna()
df=df[df['population']>=600]
df=df[df['population']<=3000]

In [ ]:
df.shape

(16271, 10)

In [ ]:
from sklearn.model_selection import train_test_split

y = df['median_house_value']
X = df[['latitude','total_rooms','median_income']]

# For the larger the dataset, the smaller the test 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20, random_state=42)

In [ ]:
X.describe()

latitude   total_rooms  median_income
count  16271.000000  16271.000000   16271.000000
mean      35.569522   2518.339623       3.867861
std        2.113046   1237.362697       1.841316
min       32.540000    121.000000       0.499900
25%       33.930000   1661.500000       2.587500
50%       34.220000   2237.000000       3.550600
75%       37.680000   3087.000000       4.728300
max       41.950000  19107.000000      15.000100

In [ ]:
from feature_engine.outliers import Winsorizer
capper = Winsorizer(capping_method='iqr', tail='both')
capper.fit(X_train)
X_train=capper.transform(X_train)
X_test=capper.transform(X_test)

In [ ]:
from sklearn.preprocessing import RobustScaler
from feature_engine.wrappers import SklearnTransformerWrapper
scaler = SklearnTransformerWrapper(transformer = RobustScaler())
scaler.fit(X_train)
X_train=scaler.transform(X_train)
X_test=scaler.transform(X_test)

In [ ]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train,y_train)




LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [ ]:
pred=model.predict(X_test)

In [ ]:
from sklearn.metrics import mean_squared_error
mean_squared_error(y_test,pred)


6071014206.6652975

In [ ]:
from sklearn.metrics import r2_score
r2_score(y_test,pred)

0.5243683694358883